Neos.ContentRepository/Classes/Domain/ValueObject/NodeName.php

```diff
@@ -0,0 +1,112 @@
+<?php
+declare(strict_types=1);
+
+namespace Neos\ContentRepository\Domain\ValueObject;
+
+/*
+ * This file is part of the Neos.ContentRepository package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\ContentRepository\Domain\Model\NodeInterface;
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * The Node Name is the "path part" of the node; i.e. when accessing the node "/foo" via path,
+ * the node name is "foo".
+ *
+ * @Flow\Proxy(false)
+ * @api
+ */
+final class NodeName implements \JsonSerializable
+{
+
+    /**
+     * @var NodeName
+     */
+    private static $rootNodeName;
+
+    /**
+     * @var NodeName
+     */
+    private static $unnamedNodeName;
+
+    /**
+     * @var string
+     */
+    private $value;
+
+    private function __construct(string $value)
+    {
+        if (!is_string($value) || preg_match(NodeInterface::MATCH_PATTERN_NAME, $value) !== 1) {
+            throw new \InvalidArgumentException('Invalid node name "' . $value . '" (a node name must only contain lowercase characters, numbers and the "-" sign).', 1364290748);
+        }
+
+        $this->value = $value;
+    }
+
+    public static function fromString(string $value): self
+    {
+        return new static($value);
+    }
+
+    /**
+     * This constructor creates an empty NodeName that can be used to create "unnamed" nodes
+     * @see isUnnamed()
+     *
+     * @return NodeName
+     */
+    public static function unnamed(): NodeName
```

Comment:
Shouldn't unnamed nodes just not have a name? Works perfectly fine...

Reply:
No opinion on this one. @bwaidelich you?

Reply:
`NodeName::unnamed()` contains some magic.. It would be nice to get around that, but I prefer it to a nullable.
So `Node::getNodeName()` should either throw an exception or we need yet another interface for _named_ nodes.
Could be discussed/solved in a separate PR IMO